### PR TITLE
Add support for exemplars during recording rule evaluation

### PR DESCRIPTION
**Pull request by vjsamuel**
This PR adds support for recording rules to emit exemplars. Would like to get some early feedback before I add things like unit test cases, feature flags (if necessary).

I wanted to add a sampling ratio to only append a subset of exemplars along with the recorded metric but would love to get feedback before hand.

### CODE REVIEW

**Comment by bboreham** [comment]
Thanks for this PR.  Could you please write some words about the requirements you intended to meet?
E.g. for a rule which outputs fewer series than it inputs, how are exemplars mapped from input to output series.
Perhaps you can add this to #8798.

It's best if we can debate the requirements separately from the implementation.

I also fear the implementation will be expensive.  One or two benchmarks might help to answer this.

**Comment by vjsamuel** [comment]
@bboreham thanks for getting back. I somehow missed this comment. 

The requirements that we intend to meet overall is that given a recording/alert rule, to add exemplars into the recorded metric/alert. This can either be done as append all raw exemplars into the new series that matches the labels on the new metric or to append a sample of them (0-1 sample ratio, with a sane default limit say 0.01).

This PR has done the first part of it which is the recording rule. (I needed some help to understand how best to add it into alerting rules as the implementation right now makes it not so straight forward to do the same).

The current implementation of ExemplarAppender expects the exemplar to be appended only into a head series that already exists if not it fails. To be true to the current requirements, the exemplars from the raw series are matched against the labels of the new recorded metric and then added into the recorded series. That being said raw metric exemplars have raw series labels. Recorded metric exemplars have recorded series labels. 

I felt that this is a happy compromise given that it is important for us to be able to have access to the exemplars on recorded metrics even if it means that all of the raw labels arent necessarily present. 

there is one portion that i yet to be done on this PR which is to have queries be made via fanout storage instead of just local storage similar to how EngineQueryFunc does today but i wanted to get a first review in case i am totally off on the implementation :)

File: tsdb/db.go [hunk 1]
```diff
@@ -925,6 +926,11 @@ func (db *DB) Appender(ctx context.Context) storage.Appender {
 	return dbAppender{db: db, Appender: db.head.Appender(ctx)}
 }
 
+// AppendExemplar appends an exemplar into the head tsdb assuming that a series ref already exists
+func (db *DB) AppendExemplar(ref storage.SeriesRef, l labels.Labels, e exemplar.Exemplar) (storage.SeriesRef, error) {
```
Comment (codesome) [inline tsdb/db.go:930]: This is not a pattern we follow in TSDB and would be a wrong usage of the appender below. The appender needs to append and then commit to assure it is put into TSDB. We should not be assuming the underlying implementation details. Instead of a helper function like this, I suggest getting an appender from the DB, append the exemplar, and commit it. (same comment for agent/db.go)

**Review by codesome — COMMENTED** [review]
This is only a comment on the TSDB part.

**Comment by bboreham** [comment]
> E.g. for a rule which outputs fewer series than it inputs, how are exemplars mapped from input to output series.

I did not follow your answer to this question.  Please could you explain it more simply for me.

**Comment by vjsamuel** [comment]
@bboreham 
let's say there is a metric `latency` which has `pod`, `namespace`, `instance`, `le`, `http_path` and `http_verb` as labels. 

a recording rule `histogram_quantile(0.99, sum(rate(latency[5m])) by (le, http_path, http_verb, namespace)` would:

* create a rolled up metric that is the 99th percentile of latency and only record path, verb and namespace as the series labels. 
* queries all exemplars for the raw query and tries to match them against series that were newly generated. ie, an examplar of series `pod="abc", namespace="def", path="/", verb="GET", le="+Inf"` would get appended into the newly created series `path="/", namespace="def", verb="GET"`

the reason we do this is because all exemplars that get appended into the exemplar storage need to have an already existing head series entry.

does this clarify?

File: tsdb/db.go (hunk 1 above)
Comment (vjsamuel) [reply tsdb/db.go]: @codesome , i have removed this piece of code as I shouldn't even need it.

**Comment by vjsamuel** [comment]
@bboreham could you kindly take a look?

File: rules/manager.go [hunk 2]
```diff
@@ -927,18 +958,19 @@ type NotifyFunc func(ctx context.Context, expr string, alerts ...*Alert)
...
-	ForGracePeriod  time.Duration
-	ResendDelay     time.Duration
-	GroupLoader     GroupLoader
+	ExternalURL       *url.URL
+	QueryFunc         QueryFunc
+	NotifyFunc        NotifyFunc
+	Context           context.Context
+	Appendable        storage.Appendable
+	Queryable         storage.Queryable
+	ExemplarQueryable storage.ExemplarQueryable
```
Comment (bboreham) [inline rules/manager.go:967]: Should this be a function analogous to `QueryFunc`, for consistency?

File: rules/recording.go [hunk 3]
```diff
@@ -73,8 +76,60 @@ func (rule *RecordingRule) Labels() labels.Labels {
...
+		for _, l := range v.Metric {
+			if l.Name != labels.MetricName {
+				matchers = append(matchers, labels.MustNewMatcher(labels.MatchEqual, l.Name, l.Value))
+			}
+		}
+
+		for i := range ex {
+			if ok := matches(ex[i].SeriesLabels, matchers); ok {
+				ex[i].SeriesLabels = v.Metric
+			}
```
Comment (bboreham) [inline rules/recording.go:122]: What happens for exemplars that never match?

**Review by bboreham — COMMENTED** [review]
I think this is a decent effort; nothing fundamentally flawed.
Probably we will need to do something more efficient than nested loops for matching, but we can look at performance after getting it to work.

Suggest writing some tests next.

I would also be concerned about overload - if I run a rule `count by (job)({job!=""})` does it fetch and re-append every exemplar?

**Comment by vjsamuel** [comment]
@bboreham thank you for the review. i have the same concern about how things like `label_replace` would work. I will do a rebase, add test cases and address all the comments that you had.

File: rules/recording.go (hunk 3 above)
Comment (vjsamuel) [reply rules/recording.go:122]: modified the code so that we drop them on the floor.

File: rules/recording.go [hunk 4]
```diff
@@ -73,8 +76,60 @@ func (rule *RecordingRule) Labels() labels.Labels {
...
+	// Query all the raw exemplars that match the query
+	ex, err := eq.Select(timestamp.FromTime(ts.Add(-1*interval)), timestamp.FromTime(ts), selectors...)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// For each vector received, we create matchers for all labels of the vector minus the metric name. If it matches
+	// we replace the exemplar's series labels with the vector's labels. We do this to ensure that we are able to append
+	// the exemplars to the same series ref the recorded metric went into.
+	for _, v := range vector {
```
[the remark this answers is not in the document]
Comment (vjsamuel) [reply rules/recording.go:111]: i am assuming that your comment is regarding the comments alone. could you kindly clarify?

File: rules/recording.go [hunk 5]
```diff
@@ -73,8 +76,60 @@ func (rule *RecordingRule) Labels() labels.Labels {
...
+	}
+
+	rule.SetHealth(HealthGood)
+	rule.SetLastError(err)
+	return vector, nil
+}
+
+// EvalWithExemplars evaluates the rule and then overrides the metric names and labels accordingly along with emitting exemplars.
+func (rule *RecordingRule) EvalWithExemplars(ctx context.Context, ts time.Time, interval time.Duration,
+	query QueryFunc, eq storage.ExemplarQuerier, _ *url.URL, limit int) (promql.Vector, []exemplar.QueryResult, error) {
```
[the remark this answers is not in the document]
Comment (vjsamuel) [reply rules/recording.go:99]: the interface is shared with the alerting implementation. if i remove the url, then it will break the flow of alerts when exemplars are turned on.

**Comment by vjsamuel** [comment]
> if I run a rule count by (job)({job!=""}) does it fetch and re-append every exemplar?
unfortunately yes. i have contemplated on adding a sampling ratio (0-1) to only add a few into the recorded metric unless the end user absolutely opts for 100% of the exemplars to be re-appended. do you think this is something that i should do?

>  what happens if the query does a label_replace so the label names don't match?
if i am not wrong, they would work fine because i am only doing:
```
selectors := parser.ExtractSelectors(rule.vector)
	if len(selectors) < 1 {
		return nil, nil, err
	}
```

once the exemplar is available, i overwrite all the labels with the labels from the vector which would take care of label_replace. please correct me if my assumption is invalid.

**Comment by vjsamuel** [comment]
@bboreham , added some test cases and realized that the promql test utility doesnt support loading exemplars :) is there a simpler way to test this feature or do i figure out how to enhance the utility?

(hope you had a nice KubeCon)

File: rules/alerting.go [hunk 6]
```diff
@@ -471,6 +472,17 @@ func (r *AlertingRule) Eval(ctx context.Context, ts time.Time, query QueryFunc,
 	return vec, nil
 }
 
+// EvalWithExemplars does everything that Eval does along with emit exemplars.
+// Right now it doesn't emit exemplars given the fact that alert state label isn't as straight forward to handle.
```
Comment (bboreham) [inline rules/alerting.go:476]: Please clear up this contradiction.

File: rules/manager.go [hunk 7]
```diff
@@ -212,21 +215,45 @@ func EngineQueryFunc(engine *promql.Engine, q storage.Queryable) QueryFunc {
...
 // interval and acted upon (currently either recorded or used for alerting).
 type Rule interface {
 	Name() string
 	// Labels of the rule.
 	Labels() labels.Labels
-	// eval evaluates the rule, including any associated recording or alerting actions.
+	// Eval evaluates the rule, including any associated recording or alerting actions.
 	Eval(context.Context, time.Time, QueryFunc, *url.URL, int) (promql.Vector, error)
+	// EvalWithExemplars evaluates the rule, including any associated recording or alerting actions and emit exemplars.
+	EvalWithExemplars(ctx context.Context, ts time.Time, interval time.Duration, query QueryFunc, exquery ExemplarQueryFunc, url *url.URL, limit int) (promql.Vector, []exemplar.QueryResult, error)
```
Comment (bboreham) [inline rules/manager.go:249]: I think it would be good to state the meaning of the parameters, at least `url`, `interval` and `limit`.
`exquery` is not Go-style naming; what about exemplarQuery ?

**Review by bboreham — COMMENTED** [review]
Thanks; code seems broadly ok.  I have some suggestions to improve performance but let's get the tests done first.

> realized that the promql test utility doesnt support loading exemplars :) is there a simpler way to test this feature or do i figure out how to enhance the utility?

Enhancing the utility sounds like the best approach.
Possibly this will be complicated, because it uses a yacc parser.
Let me know if you get stuck; I should be able to help.

**Comment by roidelapluie** [comment]
hello @vjsamuel, are you still interested to go on with this pull request and address @bboreham's comments?

Thanks!

**Comment by vjsamuel** [comment]
apologies for the massive delay @roidelapluie @bboreham, i got pulled into some internal priorities and will restart work on this in a couple of days. I will have it for review by next week.

File: rules/recording_test.go [hunk 8]
```diff
@@ -136,12 +136,56 @@ func TestRuleEval(t *testing.T) {
...
+					ExemplarQuerierQueryFunc(suite.ExemplarQueryable()), nil, 0)
+				if err != nil {
+					require.NoError(b, err)
+				}
+			}
+		})
+	}
+}
+
+func BenchmarkRuleEvalWithZeroExemplars(b *testing.B) {
```
[the remark this answers is not in the document]
Comment (vjsamuel) [reply rules/recording_test.go:183]: there is another `BenchmarkRuleEval` that calls `Eval()`. thats why i used this.

File: rules/manager.go (hunk 7 above)
Comment (vjsamuel) [reply rules/manager.go]: i have updated the above to be consistent with `Eval`. if you can please give me an example of how to document each param, i would be happy to add in the comments.

File: rules/alerting.go (hunk 6 above)
Comment (vjsamuel) [reply rules/alerting.go]: done

File: rules/recording.go [hunk 9]
```diff
@@ -71,45 +72,137 @@ func (rule *RecordingRule) Labels() labels.Labels {
...
+
+	// Query all the raw exemplars that match the query
+	ex, err := eq(ctx, rule.vector, ts, interval)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// For each vector received, we create matchers for all labels of the vector. If it matches
+	// we replace the exemplar's series labels with the vector's labels. We do this to ensure that we are able to append
+	// the exemplars to the same series ref the recorded metric went into.
```
[the remark this answers is not in the document]
Comment (vjsamuel) [reply rules/recording.go]: done

**Comment by vjsamuel** [comment]
@bboreham i have done the crudest possible hack to add in the tests. would appreciate your guidance to formally add in the parsing logic for properly loading exemplars with exemplar labels.

**Comment by vjsamuel** [comment]
*This is the new code when exemplars are turned on but there are no exemplars*
```
BenchmarkRuleEvalWithNoExemplars
BenchmarkRuleEvalWithNoExemplars/no_labels_in_recording_rule,_metric_name_in_query_result
BenchmarkRuleEvalWithNoExemplars/no_labels_in_recording_rule,_metric_name_in_query_result-10         	  126202	      9662 ns/op
BenchmarkRuleEvalWithNoExemplars/only_new_labels_in_recording_rule,_metric_name_in_query_result
BenchmarkRuleEvalWithNoExemplars/only_new_labels_in_recording_rule,_metric_name_in_query_result-10   	  126406	      9486 ns/op
BenchmarkRuleEvalWithNoExemplars/some_replacement_labels_in_recording_rule,_metric_name_in_query_result
BenchmarkRuleEvalWithNoExemplars/some_replacement_labels_in_recording_rule,_metric_name_in_query_result-10         	  125504	      9517 ns/op
BenchmarkRuleEvalWithNoExemplars/no_labels_in_recording_rule,_no_metric_name_in_query_result
BenchmarkRuleEvalWithNoExemplars/no_labels_in_recording_rule,_no_metric_name_in_query_result-10                    	   68275	     17202 ns/op
```

*This is when the feature is turned on and has exemplars*
```
BenchmarkRuleEvalWithExemplars
BenchmarkRuleEvalWithExemplars/no_labels_in_recording_rule,_metric_name_in_query_result
BenchmarkRuleEvalWithExemplars/no_labels_in_recording_rule,_metric_name_in_query_result-10                         	  114679	     10341 ns/op
BenchmarkRuleEvalWithExemplars/only_new_labels_in_recording_rule,_metric_name_in_query_result
BenchmarkRuleEvalWithExemplars/only_new_labels_in_recording_rule,_metric_name_in_query_result-10                   	  113203	     10366 ns/op
BenchmarkRuleEvalWithExemplars/some_replacement_labels_in_recording_rule,_metric_name_in_query_result
BenchmarkRuleEvalWithExemplars/some_replacement_labels_in_recording_rule,_metric_name_in_query_result-10           	  114609	     10297 ns/op
BenchmarkRuleEvalWithExemplars/no_labels_in_recording_rule,_no_metric_name_in_query_result
BenchmarkRuleEvalWithExemplars/no_labels_in_recording_rule,_no_metric_name_in_query_result-10                      	   66717	     18087 ns/op
```

*This is the current baseline*
```
BenchmarkRuleEval
BenchmarkRuleEval/no_labels_in_recording_rule,_metric_name_in_query_result
BenchmarkRuleEval/no_labels_in_recording_rule,_metric_name_in_query_result-10                                      	  136940	      8617 ns/op
BenchmarkRuleEval/only_new_labels_in_recording_rule,_metric_name_in_query_result
BenchmarkRuleEval/only_new_labels_in_recording_rule,_metric_name_in_query_result-10                                	  137937	      8669 ns/op
BenchmarkRuleEval/some_replacement_labels_in_recording_rule,_metric_name_in_query_result
BenchmarkRuleEval/some_replacement_labels_in_recording_rule,_metric_name_in_query_result-10                        	  137877	      8624 ns/op
BenchmarkRuleEval/no_labels_in_recording_rule,_no_metric_name_in_query_result
BenchmarkRuleEval/no_labels_in_recording_rule,_no_metric_name_in_query_result-10                                   	   73915	     16336 ns/op
PASS
```

There seems to be a 7% slow down from the baseline when turning on the feature even when there is no exemplars.

**Comment by bboreham** [comment]
Tip: run your benchmarks several times to account for random slow-downs on your computer. For example `-count=6`.
Then use the `benchstat` tool to analyze the timings and give you the difference and variance.
Install instructions here: https://cs.opensource.google/go/x/perf

File: rules/recording_test.go [hunk 10]
```diff
@@ -120,6 +120,21 @@ func setUpRuleEvalTest(t require.TestingT) *promql.Test {
...
 
+func setUpRuleEvalExemplarTest(t require.TestingT) *promql.Test {
+	suite, err := promql.NewTest(t, `
+		load 1m
+			metric{label_a="1",label_b="3"} 1
+			metric{label_a="2",label_b="4"} 10
+
+		load_exemplars 1m
+			metric{label_a="1",label_b="3"} 1
+			metric{label_a="2",label_b="4"} 10
```
Comment (bboreham) [inline rules/recording_test.go:131]: Can you explain this?  I expect an exemplar to attach to an existing series and add some more labels (in practice, exactly one label "traceid" is typical).

Comment (bwplotka) [reply rules/recording_test.go:131]: Agree too, it's weird. I wonder if we could extend load to add series and some notion of exemplar somewhere?

Comment (roidelapluie) [reply rules/recording_test.go:131]: I agree with Bryan, we should work closer to the openmetrics format and use traceid as an exemplar.
`load 1m metric 1 1 1 1 # {traceid="1234"} 2 10 # {traceid="1235"} 4 10 # {traceid="1236"} 6 10`

**Comment by bwplotka** [comment]
Yup, see  https://github.com/bwplotka/go-proto-bench/blob/main/Makefile#L75 as an example framework to use (make bench and make cmp)

**Comment by krajorama** [comment]
Hello from the bug scrub @vjsamuel : we think this is  a useful feature, but some might not want to take the performance hit. We suggest the following: 
- Adding a feature flag to be able to turn the feature on, but avoid performance regression. Documentation among feature flags.
- Documentation should mention the performance impact

**Comment by fwippe** [comment]
Hey @vjsamuel @bboreham @krajorama, I would like to start working on this issue, based on the changes made as part of this PR. I will focus on recording rules, and would leave alerts alone for now.

**Comment by fwippe** [comment]
@bboreham @krajorama created PR https://github.com/prometheus/prometheus/pull/15555, picking up the work where @vjsamuel ended. I adjusted the test parser to be able to process exemplars data as part of the series definition, also adding an experimental feature flag `exemplars-in-recording-rules`. Would appreciate any feedback.

**Comment by bboreham** [comment]
Hello again from the bug-scrub!  Given no movement here, we will close this PR in favour of #15555.
